File: 33/solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        start = 0
        end = len(nums)-1
        if nums[start] == target:
            return start
        elif nums[end] == target:
            return end
        
        while end > start+1:
            mid = (start + end)/2
            if nums[mid] == target:
                return mid
            
            if nums[mid] < nums[end]:
                if nums[mid] <= target and target <= nums[end]:
                    start = mid
                else:
                    end = mid
            else:
                if nums[start] <= target and target <= nums[mid]:
                    end = mid
                else:
                    start = mid

        return -1
        

File: 33/solution.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size()-1;
        
        if(nums[left] == target)
            return left;
        else if(nums[right] == target)
            return right;
        while(right > left+1){
            if(nums[(left+right)/2] == target)
                return (left+right)/2;

            if(nums[(left+right)/2] < nums[right]){
                if(nums[(left+right)/2] < target && target < nums[right])
                    left = (left+right)/2;
                else
                    right = (left+right)/2;
            }else{
                if(nums[left] < target && target < nums[(left+right)/2])
                    right = (left+right)/2;
                else
                    left = (left+right)/2;
            }
        }
        return -1;
    }
};
